Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/cache/downloader/PartialContentSupportChecker.kt

```diff
@@ -0,0 +1,271 @@
+package com.github.adamantcheese.chan.core.cache.downloader
+
+import android.util.LruCache
+import androidx.annotation.GuardedBy
+import com.github.adamantcheese.chan.core.cache.FileCacheV2
+import com.github.adamantcheese.chan.core.cache.downloader.DownloaderUtils.isCancellationError
+import com.github.adamantcheese.chan.core.di.NetModule
+import com.github.adamantcheese.chan.utils.Logger
+import io.reactivex.Single
+import io.reactivex.SingleEmitter
+import okhttp3.*
+import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
+import java.io.IOException
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.TimeoutException
+
+/**
+ * This class is used to figure out whether an image or a file can be downloaded from the server in
+ * separate chunks concurrently using HTTP Partial-Content. For batched image downloading and
+ * media prefetching this method returns false because we should download them normally. Chunked
+ * downloading should only be used for high priority files/images like in the gallery when the user
+ * is viewing them. Everything else should be downloaded in a singe chunk.
+ * */
+internal class PartialContentSupportChecker(
+        private val okHttpClient: OkHttpClient,
+        private val activeDownloads: ActiveDownloads,
+        private val maxTimeoutMs: Long
+) {
+    // Thread safe
+    private val cachedResults = LruCache<String, PartialContentCheckResult>(1024)
+    @GuardedBy("itself")
+    private val checkedChanHosts = mutableMapOf<String, Boolean>()
+
+    fun check(url: String): Single<PartialContentCheckResult> {
+        if (activeDownloads.isBatchDownload(url)) {
+            return Single.just(PartialContentCheckResult(false))
+        }
+
+        val fileSize = activeDownloads.get(url)?.extraInfo?.fileSize ?: -1L
+        if (fileSize > 0) {
+            val host = url.toHttpUrlOrNull()?.host
+
+            if (host != null) {
+                val supportsPartialContent = synchronized(checkedChanHosts) {
+                    checkedChanHosts[host] ?: false
+                }
+
+                if (supportsPartialContent) {
```

Comment:
I think I forgot to handle the case when supportsPartialContent == false. In such case we don't want to send HEAD request because we already know that this site does not support PartialContent - there is no point. So we should return immediately as well with supportsPartialContentDownload == false.